### PR TITLE
build: use LuaRocks 2.4.2

### PR DESCRIPTION
**Pull request by javian**
#5699 was never completed and merged and the issue never stated why so I'm going to give this a go again.

### CODE REVIEW

**Comment by javian** [comment]
Looks like they split up the windows at unix source archive from 2.3 and onwards. I'll see if I can make some edits to make it work.

**Comment by justinmk** [comment]
You mean this commit? https://github.com/luarocks/luarocks/commit/1dd5293f0c725c1d5f290bbed51b3bec59748d16

Thanks for looking into this, we definitely need it.

**Comment by javian** [comment]
Ah yes, that's what I was wanting to look for. So it looks like some paths needs to change which shouldn't be undoable =) Would you have any tips on how to test this for the windows build @justinmk if I wanted to do it locally rather than relying on appveyor ? I'm using osx.

**Comment by javian** [comment]
Actually it looks like the commit you posted only references the luarocks repo internally so I don't think it affects anything with regards to the upgraded version.

**Comment by javian** [comment]
The error on windows currently comes down to
```
-- Could NOT find LIBUV (missing:  LIBUV_INCLUDE_DIR) 
-- Could NOT find LUAJIT (missing:  LUAJIT_LIBRARIES LUAJIT_INCLUDE_DIR) 
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
LIBUV_LIBRARIES
    linked by target "luv" in directory C:/projects/neovim/.deps/build/src/luv
LUAJIT_LIBRARIES
    linked by target "luv" in directory C:/projects/neovim/.deps/build/src/luv
```

**Comment by javian** [comment]
The cause lies in the compiled/generated luarocks.bat. It contains a non-working path to luajit.exe and the problem was introduced somewhere between the luarocks release of https://github.com/luarocks/luarocks/releases/tag/v2.2.2 and https://github.com/luarocks/luarocks/releases/tag/v2.2.3-rc1 . Still digging.

**Comment by javian** [comment]
I've not been able to get any further with this yet but I will keep looking.

**Comment by javian** [comment]
I think I managed to figure it out and it currently goes through the third party build step without issues locally on my machine. The appveyor CI seems to fail for other reasons and I couldn't understand why the Travis errors ended up with this
```
[ 80%] Performing install step for 'luarocks'
/home/travis/nvim-deps/usr/bin/luajit: ...travis/nvim-deps/build/src/luarocks/src/luarocks/cfg.lua:78: attempt to index local 'site_config' (a boolean value)
stack traceback:
	...travis/nvim-deps/build/src/luarocks/src/luarocks/cfg.lua:78: in main chunk
	[C]: in function 'require'
	src/bin/luarocks:4: in main chunk
	[C]: at 0x00404b20
make[3]: *** [run_luarocks] Error 1
make[3]: *** Waiting for unfinished jobs....
make[2]: *** [build/src/luarocks-stamp/luarocks-install] Error 2
make[1]: *** [CMakeFiles/luarocks.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by justinmk** [comment]
deleted the travis cache for this PR, just in case.

**Comment by javian** [comment]
@justinmk Do I need to rebase this PR against master ?

**Comment by justinmk** [comment]
That might help. The appveyor build is failing to download libvterm, so something's messed up.

**Comment by javian** [comment]
Ok that seemed to to do the trick. Got a green light on 2 out of 3 and Travis failed for unrelated reasons. I squashed the commit now and its running the ci again.

**Comment by javian** [comment]
@justinmk any ideas ? Travis still seems to be failing for reasons not related to the code.

**Comment by justinmk** [comment]
@javian Restarted the build. Might also try removing `ubuntu-toolchain-r/test` from .travis.yml, I suspect it isn't needed since we moved to "trusty" travis env.

File: third-party/cmake/BuildLuarocks.cmake [hunk 1]
```diff
@@ -73,13 +73,13 @@ elseif(MSVC OR MINGW)
...
     /INC ${DEPS_INSTALL_DIR}/include/luajit-2.0/
-    /P ${DEPS_INSTALL_DIR} /TREE ${DEPS_INSTALL_DIR}
+    /P ${DEPS_INSTALL_DIR}/2.4 /TREE ${DEPS_INSTALL_DIR}
     /SCRIPTS ${DEPS_BIN_DIR}
     /CMOD ${DEPS_BIN_DIR}
     ${MINGW_FLAG}
     /LUAMOD ${DEPS_BIN_DIR}/lua)
 
-  set(LUAROCKS_BINARY ${DEPS_INSTALL_DIR}/2.2/luarocks.bat)
+  set(LUAROCKS_BINARY ${DEPS_INSTALL_DIR}/2.4/luarocks.bat)
```
Comment (justinmk) [inline third-party/cmake/BuildLuarocks.cmake]: Let's put the 2.4 into a variable (say after line 42):

    set(LUAROCKS_VERSION 2.4)

File: third-party/CMakeLists.txt [hunk 2]
```diff
@@ -99,8 +99,8 @@ set(LUAJIT_SHA256 620fa4eb12375021bef6e4f237cbd2dd5d49e56beb414bee052c746beef180
 set(LUA_URL https://www.lua.org/ftp/lua-5.1.5.tar.gz)
 set(LUA_SHA256 2640fc56a795f29d28ef15e13c34a47e223960b0240e8cb0a82d9b0738695333)
 
-set(LUAROCKS_URL https://github.com/keplerproject/luarocks/archive/5d8a16526573b36d5b22aa74866120c998466697.tar.gz)
-set(LUAROCKS_SHA256 cae709111c5701235770047dfd7169f66b82ae1c7b9b79207f9df0afb722bfd9)
+set(LUAROCKS_URL https://github.com/luarocks/luarocks/archive/v2.4.2.tar.gz)
+set(LUAROCKS_SHA256 eef88c2429c715a7beb921e4b1ba571dddb7c74a250fbb0d3cc0d4be7a5865d9)
```
Comment (justinmk) [inline third-party/CMakeLists.txt]: Add a comment above this so we don't have to rediscover this again :)

    # NOTE: Version must match LUAROCKS_VERSION in BuildLuarocks.cmake

**Comment by jamessan** [comment]
> Might also try removing ubuntu-toolchain-r/test from .travis.yml

It is still needed.

**Comment by javian** [comment]
@justinmk I'll make the updates. Travis seems to be doing better/fine now and it just timed out due to the build taking too long.

File: third-party/cmake/BuildLuarocks.cmake (hunk 1 above)
Comment (javian) [inline third-party/cmake/BuildLuarocks.cmake]: I wonder if we really need to use the version number or if we just could use a generic term instead. The documentation says
```
/P [dir]       Where to install LuaRocks. 
               Default is %PROGRAMFILES%\LuaRocks
```
so I could test and replace it.

File: third-party/cmake/BuildLuarocks.cmake [hunk 3]
```diff
@@ -73,13 +76,13 @@ elseif(MSVC OR MINGW)
...
     /INC ${DEPS_INSTALL_DIR}/include/luajit-2.0/
-    /P ${DEPS_INSTALL_DIR} /TREE ${DEPS_INSTALL_DIR}
+    /P ${DEPS_INSTALL_DIR}/${LUAROCKS_VERSION} /TREE ${DEPS_INSTALL_DIR}
     /SCRIPTS ${DEPS_BIN_DIR}
     /CMOD ${DEPS_BIN_DIR}
     ${MINGW_FLAG}
     /LUAMOD ${DEPS_BIN_DIR}/lua)
 
-  set(LUAROCKS_BINARY ${DEPS_INSTALL_DIR}/2.2/luarocks.bat)
+  set(LUAROCKS_BINARY ${DEPS_INSTALL_DIR}/${LUAROCKS_VERSION}/luarocks.bat)
```
Comment (justinmk) [inline third-party/cmake/BuildLuarocks.cmake]: I wonder if it works if you _remove_ the version from both here and line 79?

**Comment by justinmk** [comment]
@javian Those aren't timeouts, it's failing at the "install" step:

```
[ 88%] Performing install step for 'luarocks'
/home/travis/nvim-deps/usr/bin/luajit: ...travis/nvim-deps/build/src/luarocks/src/luarocks/cfg.lua:78: attempt to index local 'site_config' (a boolean value)
stack traceback:
	...travis/nvim-deps/build/src/luarocks/src/luarocks/cfg.lua:78: in main chunk
	[C]: in function 'require'
	src/bin/luarocks:4: in main chunk
	[C]: at 0x00404b20
```

I noticed that your change is for the `(MSVC OR MINGW)` case, so it doesn't affect travis. Not sure why it works on quickbuild. Does it work for you locally (assuming you run linux/bsd) after `make distclean`?

File: third-party/cmake/BuildLuarocks.cmake (hunk 3 above)
Comment (javian) [inline third-party/cmake/BuildLuarocks.cmake]: That is what the old PR did - https://github.com/neovim/neovim/pull/5699/files which didn't work.

Comment (justinmk) [inline third-party/cmake/BuildLuarocks.cmake]: Ok. Really weird that travis _does_ work in that PR. Perhaps try using v2.4.0 instead of v2.4.2?

**Comment by javian** [comment]
It works fine on macOS and testing it on my ubuntu machine now (16.04) and it seems to be doing just fine and doesn't fail in the same place. The first travis job is the one that times out - but the other ones are currently failing like you describe.

**Comment by justinmk** [comment]
@javian The travis logs suggest that cache is being used, but I don't see an item for this PR in the "Delete cache" UI in travis. Could you try opening a new PR (use a different branch name on your fork, just in case).

**Comment by javian** [comment]
Done - lets see how it goes.

**Comment by javian** [comment]
Looks like its failing in the same place on the other PR as well.

File: third-party/cmake/BuildLuarocks.cmake (hunk 3 above)
Comment (javian) [inline third-party/cmake/BuildLuarocks.cmake]: Testing it now.

**Comment by javian** [comment]
2.4.0 seems to have worked https://travis-ci.org/neovim/neovim/builds/227094405 - testing 2.4.1 next.

**Comment by javian** [comment]
and 2.4.1 fails

**Comment by justinmk** [comment]
Ok great, now we can look at luarocks repo history to see what changed.

**Comment by javian** [comment]
https://github.com/luarocks/luarocks/compare/2.4.0...2.4.1 luarocks diff. There's seems to have been some change made to site_config related code

**Comment by javian** [comment]
I can see that the different travis builds use different deps directories. One stems from the cache and the other one doesn't as far as I can tell now but I haven't been able to figure out why one fails and not the other. Do you think it would be an option to test to build the third part deps on travis ie. set `BUILD_NVIM_DEPS=true` @justinmk ?

I also noticed that somewhere along the lines the second MacOS test is now successful where it used to fail so not sure what has happened there.

**Comment by justinmk** [comment]
> Do you think it would be an option to test to build the third part deps on travis ie. set BUILD_NVIM_DEPS=true @justinmk ?

@javian Yes, feel free to set any variables you need in `.travis.yml`.

**Comment by justinmk** [comment]
Nice. So it's working with 2.4.1 and BUILD_NVIM_DEPS=true. Maybe 2.4.2 will work? Though I don't see why BUILD_NVIM_DEPS should matter, especially since we deleted the travis cache for this PR.

**Comment by javian** [comment]
Is this really correct code ? https://github.com/neovim/neovim/blob/master/ci/common/build.sh#L38 if it isn't then the third party software would never get update unless forced.

**Comment by fwalch** [comment]
> since we deleted the travis cache for this PR

If the PR cache is deleted, Travis just takes the one from `master`.

> maybe our caching logic is not automatically invalidating itself when changes are made to third-party/*

From a quick look, the line @javian posted should work, but honestly I can't remember the details, it's been quite a while.. so to be safe: temporarily set `BUILD_NVIM_DEPS=true` in the PR to test the new deps, then set back to `false`, merge the PR and delete *all* Travis caches.

**Comment by javian** [comment]
The current travis failure doesn't seem to be related to the PR itself.

**Comment by justinmk** [comment]
@javian Are the latest commits just some final sanity checks? This was working as far as we know, right?

**Comment by javian** [comment]
@justinmk yeah, as far as I could tell but I believe as far as I can recall it worked slightly better with an earlier version so I was aiming to hit green all over. Not sure if that can be done though.

**Comment by javian** [comment]
@justinmk I also just noticed that there was a failing test on the last build - https://ci.appveyor.com/project/neovim/neovim/build/13530/job/cqksu610xo0n0969 . I can't judge the importance of it though but it looked suspicious.

**Comment by justinmk** [comment]
The builds from https://github.com/neovim/neovim/pull/6059/commits/a435c3219923bfc7b7cd52b42f9f365e48ae5643 look fine, the failures aren't related to this PR.

I suggest rebasing https://github.com/neovim/neovim/pull/6059/commits/a435c3219923bfc7b7cd52b42f9f365e48ae5643 onto master the force-push.

**Comment by javian** [comment]
@justinmk like that ?

**Comment by equal-l2** [comment]
Can someone re-run CI?

**Comment by justinmk** [comment]
@equal-l2 done

**Comment by equal-l2** [comment]
@justinmk 
Thanks, it seems to be working fine.

**Comment by javian** [comment]
Finally it is all green ^^

**Comment by equal-l2** [comment]
@justinmk Would you merge this?

**Comment by justinmk** [comment]
Merged. Thanks @javian !

**Comment by equal-l2** [comment]
Thank you for all your work, @javian !
Also thank you for merging, @justinmk!

File: third-party/cmake/BuildLuarocks.cmake (hunk 1 above)
Comment (blueyed) [reply third-party/cmake/BuildLuarocks.cmake]: Just for reference: the version is not really necessary (and confused me with https://github.com/neovim/neovim/pull/10292).
The problem with https://github.com/neovim/neovim/pull/5699 appears to have been something different likely.

**Pull request closed without merging**